Add Liquity LQTY staking strategy

georgeroman opened the pull request:
LQTY stakers on Liquity earn pro-rata shares of borrowing and redemption fees in LUSD and ETH (https://docs.liquity.org/faq/staking). At the moment the APR on LQTY staking sits at around 10%.

Below is a run of the tests. Since rewards depend on parties borrowing and redeeming LUSD, I setup the tests to only run for a few loops, mimicking real borrowing and redeeming transactions that generate fees in both LUSD and ETH to the stakers.

```bash
  LQTY Staking
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0xF6323aD07c30D696D3e572cb4648814F1188372D
0x00000000a4569cda123E73F94e35f9a625650Be6
New Vault Deployed:  0xCa9accDBc83566610F56544B9F2210779F173bEc
Strategy Deployed:  0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
Strategy and vault added to Controller.
    Happy path
loop 0
old shareprice: 1000000000000000000
new shareprice: 1000000000000000000
growth: 1
loop 1
old shareprice: 1000000000000000000
new shareprice: 1000025547148501574
growth: 1.0000255471485016
loop 2
old shareprice: 1000025547148501574
new shareprice: 1000030407729851896
growth: 1.0000048604571794
Earned 745021573031375568
      ✓ Farmer should earn money (2361ms)


  1 passing (13s)
```

breadforthepeople commented:
Please use the pattern that is applied in the repo: 
1. Creating a core strategy that does not have the specific contract addresses
2. Having a "*Mainnet" strategy that is based on the core strategy and substitutes the addresses.
Example: https://github.com/harvest-finance/harvest-strategy/tree/master/contracts/strategies/idle

breadforthepeople commented on contracts/strategies/liquity/LiquityLQTYStakingStrategy.sol:
```diff
@@ -0,0 +1,311 @@
+     *      governance can still trigger an emergency exit, in order to
+     *      pause investing in the pool.
+     */
+    function emergencyExit() public onlyGovernance {
```
This also needs to withdraw from the reward pool.

breadforthepeople commented on contracts/strategies/liquity/LiquityLQTYStakingStrategy.sol:
```diff
@@ -0,0 +1,311 @@
+
+        // At this point, both LUSD and ETH rewards were swapped for USDC,
+        // which we now handle as the reward token
+        uint256 rewardBalance = IERC20(rewardToken()).balanceOf(address(this));
```
Could we use WETH as a rewardToken() instead? This would reduce the number of hops, and it's what we commonly did for cases when there are multiple liquidations

breadforthepeople commented on contracts/strategies/liquity/LiquityLQTYStakingStrategy.sol:
```diff
@@ -0,0 +1,311 @@
+
+        // Liquidate the ETH rewards to USDC
+        if (ethBalance > 0) {
+            ISwapRouter(uniswapV3Router).exactInputSingle.value(ethBalance)(
```
Could we give the universal liquidator a try? This has a more convenient interface. The id for the dex is "uniV3".

https://github.com/harvest-finance/harvest-strategy/blob/master/contracts/strategies/balancer/BalancerStrategy.sol#L111?

georgeroman replied:
Is the code for the UniswapV3 DEX contract public (eg. https://etherscan.io/address/0x1D35ba854575B576B3C0aB4e64E27Bf2D2c1D48A)? Not sure how to specify the path since the UniversalLiquidator contract (eg. https://etherscan.io/address/0x875680a120597732f92bf649cacfeb308e54dba4) accepts an array of addresses as the path while UniswapV3 deals with the path as bytes?

breadforthepeople replied:
Some examples are in our `FeeRewardForwarder` that swaps a reward token into FARM: https://etherscan.io/address/0x153C544f72329c1ba521DDf5086cf2fA98C86676#code
```
    storedLiquidationPaths[sushi][farm] = [sushi, weth, farm];
    storedLiquidationDexes[sushi][farm] = [sushiDex, uniDex];
```
- This basically means that `SUSHI` is liquidated into `WETH` using `Sushiswap` (`sushiDex`), and `WETH` liquidated into `FARM` using `Uniswap` (`uniDex`). Dexes are bytes32 which are defined as:

```
  bytes32 public uniDex = bytes32(uint256(keccak256("uni")));
  bytes32 public sushiDex = bytes32(uint256(keccak256("sushi")));
```

and for uniswap V3, you could use:

```
bytes32 public uniV3Dex = bytes32(uint256(keccak256("uniV3")));
```

georgeroman replied:
Thanks, that was very clear!

georgeroman commented:
Addressed the review comments. Here are the new test results:

```bash
  LQTY Staking
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0xF6323aD07c30D696D3e572cb4648814F1188372D
0x00000000a4569cda123E73F94e35f9a625650Be6
New Vault Deployed:  0xCa9accDBc83566610F56544B9F2210779F173bEc
Strategy Deployed:  0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
Strategy and vault added to Controller.
    Happy path
loop 0
old shareprice: 1000000000000000000
new shareprice: 1000000000000000000
growth: 1
loop 1
old shareprice: 1000000000000000000
new shareprice: 1000025572716378751
growth: 1.0000255727163787
loop 2
old shareprice: 1000025572716378751
new shareprice: 1000030412243389919
growth: 1.0000048394032546
Earned 745132159491254272
      ✓ Farmer should earn money (2277ms)


  1 passing (25s)
```

georgeroman commented on contracts/strategies/liquity/LQTYStakingStrategy.sol:
```diff
@@ -0,0 +1,312 @@
+            // as that's where most of the LUSD liquidity resides
+            IERC20(lusd).safeApprove(lusd3CrvPair, 0);
+            IERC20(lusd).safeApprove(lusd3CrvPair, lusdBalance);
+            ICurveMetapool(lusd3CrvPair).exchange_underlying(
```
I might be wrong but Curve doesn't seem to be supported by the Universal Liquidator yet. If it were, this swap could be made via the UL instead.

georgeroman commented on test/liquity/lqty-staking.js:
```diff
@@ -0,0 +1,194 @@
+    });
+  };
+
+  const setupULPath = async () => {
```
The WETH -> LQTY path is available in the UL registry, so it needs to be added before deploying the strategy.

CryptJS13 replied:
Hey, just FYI, the logic for adding liquidation paths is included in `hh-utils.js`. You can use it by adding a `liquidation` argument to the `setupCoreProtocol` function inputs. In this case the line would be:

`liquidation: [{ "uniV3": ["0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2", "0x6DEA81C8171D0bA574754EF6F8b412F2Ed88c54D"]}]`

The liquidation line with the weth-farm path is unnecessary because that path is already stored in the UL. If you would want to add multiple paths in one run it would look like:

`liquidation: [{ "uni": [weth, farm] }, { "uniV3": ["0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2", "0x6DEA81C8171D0bA574754EF6F8b412F2Ed88c54D"]}]`

georgeroman replied:
Thanks for the tip!

georgeroman commented:
Thanks for the tip @CryptJS13!

On Fri, 23 Jul 2021 at 12:16, CryptJS13 ***@***.***> wrote:

> *@CryptJS13* commented on this pull request.
> ------------------------------
>
> In test/liquity/lqty-staking.js
> <https://github.com/harvest-finance/harvest-strategy/pull/39#discussion_r675420439>
> :
>
> > +    );
> +    troveManager = await TroveManager.at(
> +      "0xA39739EF8b0231DbFA0DcdA07d7e29faAbCf4bb2"
> +    );
> +  };
> +
> +  const setupBalance = async () => {
> +    farmerBalance = await underlying.balanceOf(underlyingWhale);
> +
> +    // Transfer underlying from whale to farmer
> +    await underlying.transfer(farmer, farmerBalance, {
> +      from: underlyingWhale,
> +    });
> +  };
> +
> +  const setupULPath = async () => {
>
> Hey, just FYI, the logic for adding liquidation paths is included in
> hh-utils.js. You can use it by adding a liquidation argument to the
> setupCoreProtocol function inputs. In this case the line would be:
>
> liquidation: [{ "uniV3": ["0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
> "0x6DEA81C8171D0bA574754EF6F8b412F2Ed88c54D"]}]
>
> The liquidation line with the weth-farm path is unnecessary because that
> path is already stored in the UL. If you would want to add multiple paths
> in one run it would look like:
>
> liquidation: [{ "uni": [weth, farm] }, { "uniV3":
> ["0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
> "0x6DEA81C8171D0bA574754EF6F8b412F2Ed88c54D"]}]
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/harvest-finance/harvest-strategy/pull/39#discussion_r675420439>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHKY5T5D62JPZKKSDB3CWETTZEXP7ANCNFSM5ASIDKCA>
> .
>
